File: day04/camp_cleanup.py
from sys import argv

def contains(elf):
    if elf[0][0] >= elf[1][0] and elf[0][1] <= elf[1][1]:
        return True
    elif elf[1][0] >= elf[0][0] and elf[1][1] <= elf[0][1]:
        return True
    else:
        return False

def overlap(elf):
    if elf[1][0] <= elf[0][0] <= elf[1][1]:
        return True
    elif elf[1][0] <= elf[0][1] <= elf[1][1]:
        return True
    elif elf[0][0] <= elf[1][0] <= elf[0][1]:
        return True
    elif elf[0][0] <= elf[1][1] <= elf[0][1]:
        return True
    else:
        return False

elves = []
with open(argv[1]) as f:
    for line in f:
        toks = line.rstrip().split(',')
        pair = []
        for tok in toks:
            pair.append([int(x) for x in tok.split('-')])
        elves.append(pair)

contains_count = 0
overlap_count = 0
for elf in elves:
    if contains(elf):
        contains_count += 1
    if overlap(elf):
        overlap_count += 1

print('Part 1:', contains_count)
print('Part 2:', overlap_count)

File: day04/camp_cleanup.pl
#!/usr/bin/env perl
use v5.36;

sub contains(@elf) {
    if ($elf[0] >= $elf[2] && $elf[1] <= $elf[3]) {
        return 1;
    } elsif ($elf[2] >= $elf[0] && $elf[3] <= $elf[1]) {
        return 1;
    } else {
        return 0;
    }
}

sub overlap(@elf) {
    if ($elf[2] <= $elf[0] <= $elf[3]) {
        return 1;
    } elsif ($elf[2] <= $elf[1] <= $elf[3]) {
        return 1;
    } elsif ($elf[0] <= $elf[2] <= $elf[1]) {
        return 1;
    } elsif ($elf[0] <= $elf[3] <= $elf[1]) {
        return 1;
    } else {
        return 0;
    }
}

my @elves;
while (<>) {
    chomp;
    push @elves, [split /[,\-]/];
}

my $contains_count;
my $overlap_count;
for my $elf (@elves) {
    $contains_count++ if contains(@$elf);
    $overlap_count++ if overlap(@$elf);
}

say "Part 1: $contains_count";
say "Part 2: $overlap_count";
